File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/data/database/UserRealmDataStore.java
package com.lattis.ellipse.data.database;

import com.lattis.ellipse.data.database.base.RealmObservable;
import com.lattis.ellipse.data.database.mapper.RealmUserMapper;
import com.lattis.ellipse.data.database.model.RealmUser;
import com.lattis.ellipse.domain.model.User;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.functions.Function;
import io.realm.Realm;
import io.realm.RealmConfiguration;


public class UserRealmDataStore {

    private RealmConfiguration realmConfiguration;
    private RealmUserMapper userMapper;

    @Inject
    public UserRealmDataStore(RealmConfiguration realmConfiguration,
                              RealmUserMapper userMapper) {
        this.realmConfiguration = realmConfiguration;
        this.userMapper = userMapper;
    }

    public Observable<User> getUser() {
        return RealmObservable.object(
                realm -> {
                    RealmUser realmUser = realm.where(RealmUser.class)
                        .findFirst();
                    return realmUser!=null ? realmUser : new RealmUser();
                })
                .map(realmUser -> userMapper.mapOut(realmUser));
    }

    public Observable<User> createOrUpdateUser(User user) {
        return RealmObservable.object(
                new Function<Realm, RealmUser>() {
                    @Override
                    public RealmUser apply(Realm realm) {
                        return realm.copyToRealmOrUpdate(userMapper.mapIn(user));
                    }
                })
                .map(new Function<RealmUser, User>() {
                    @Override
                    public User apply(RealmUser realmUser) {
                        return userMapper.mapOut(realmUser);
                    }
                });
    }
}

File: velolabs_repos/lattis_android_v2/data/src/main/java/com/lattis/data/database/store/UserRealmDataStore.kt
package com.lattis.data.database.store

import com.lattis.data.database.base.RealmObservable
import com.lattis.data.database.mapper.RealmUserMapper
import com.lattis.data.database.model.RealmUser
import com.lattis.domain.models.User
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.functions.Function
import io.realm.Realm
import io.realm.RealmConfiguration
import javax.inject.Inject
class UserRealmDataStore @Inject
constructor(private val userMapper: RealmUserMapper) {

    val user: Observable<User>
        get() = RealmObservable.`object`<RealmUser>(
                Function{ realm ->
                    val realmUser = realm.where(RealmUser::class.java)
                        .findFirst()
                    if (realmUser != null) realmUser else RealmUser()
                })
            .map { realmUser -> userMapper.mapOut(realmUser) }

    fun createOrUpdateUser(user: User): Observable<User> {
        return RealmObservable.`object`(
                Function<Realm, RealmUser> { realm -> realm.copyToRealmOrUpdate(userMapper.mapIn(user)) })
            .map { realmUser -> userMapper.mapOut(realmUser) }
    }
}